ssh/controlmaster.go

```diff
@@ -0,0 +1,233 @@
+// Copyright 2013 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package ssh
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+	"net"
+)
+
+// controlMasterMessage covers multiple CM packets that have no data
+type controlMasterMessage struct {
+	PacketType uint32 // MUX_C_ALIVE_CHECK|MUX_C_PROXY|MUX_S_PROXY|MUX_S_OK
+	RequestID  uint32
+}
+
+// controlMasterAliveMessage is a response to a MUX_C_ALIVE_CHECK request
+type controlMasterAliveMessage struct {
+	PacketType uint32 // MUX_S_ALIVE
+	RequestID  uint32
+	ServerPID  uint32
+}
+
+// controlMasterHelloMessage is used to ask/send the protocol version
+type controlMasterHelloMessage struct {
+	PacketType      uint32 // MUX_MSG_HELLO
+	ProtocolVersion uint32
+}
+
+// DialControlMaster starts a client connection to the given SSH server.
+// It is a convenience function that connects to the given unix socket,
+// initiates the ControlMaster proxy prottocol, and then sets up a Client.
+// For access to incoming channels and requests, use net.Dial with
+// NewControlMasterClientConn instead.
+func DialControlMaster(path string) (*Client, error) {
+	conn, err := net.Dial("unix", path)
+	if err != nil {
+		return nil, err
+	}
+	c, chans, reqs, err := NewControlMasterClientConn(conn)
+	if err != nil {
+		return nil, err
+	}
+	return NewClient(c, chans, reqs), nil
+}
+
+// NewControlMasterClientConn establishes a ControlMaster connection
+// using c as the underlying transport. The Request and NewChannel
+// channels must be serviced or the connection will hang.
+func NewControlMasterClientConn(c net.Conn) (Conn, <-chan NewChannel, <-chan *Request, error) {
+	// We never populate transport, just mux since there is no handshake needed
+	conn := &connection{
+		sshConn: sshConn{conn: c},
+	}
+
```

Comment:
You might consider using a buffered reader for the read side of the connection; it may result in fewer system calls for reading the length prefix and body.  (Not sure how much performance matters here)

You could probably also reduce allocations by using a buffered writer, so you can just write the length, type, body and then flush.